llvm/lib/SYCLLowerIR/LowerESIMD.cpp

```diff
@@ -0,0 +1,1313 @@
+        Kinds,
+        getMD(llvm::ConstantInt::getNullValue(I32Ty)), // SLM size in bytes
+        getMD(llvm::ConstantInt::getNullValue(I32Ty)), // arg offsets
+        IOKinds,
+        ArgDescs};
+
+    // Add this kernel to the root.
+    Kernels->addOperand(MDNode::get(Ctx, MDArgs));
+    F.addFnAttr("oclrt", "1");
+    F.addFnAttr("CMGenxMain");
+  }
+}
+
+// collect all the vector-types that are used by genx-volatiles
+void SYCLLowerESIMDLegacyPass::collectGenXVolatileType(Module &M) {
+  for (auto &G : M.getGlobalList()) {
+    if (!G.hasAttribute("genx_volatile"))
+      continue;
+    auto PTy = dyn_cast<PointerType>(G.getType());
+    if (!PTy)
+      continue;
+    auto GTy = dyn_cast<StructType>(PTy->getPointerElementType());
+    if (!GTy || !GTy->getName().endswith("cl::sycl::intel::gpu::simd"))
+      continue;
+    assert(GTy->getNumContainedTypes() == 1);
+    auto VTy = GTy->getContainedType(0);
+    assert(VTy->isVectorTy());
+    GenXVolatileTypeSet.insert(VTy);
+  }
+}
+
+} // namespace
+
+PreservedAnalyses SYCLLowerESIMDPass::run(Function &F,
+                                          FunctionAnalysisManager &FAM,
+                                          SmallPtrSet<Type *, 4> &GVTS) {
+  // Only consider functions marked with !sycl_explicit_simd
+  if (F.getMetadata("sycl_explicit_simd") == nullptr)
+    return PreservedAnalyses::all();
+
+  SmallVector<CallInst *, 32> ESIMDIntrCalls;
+  SmallVector<Instruction *, 8> ESIMDToErases;
+
+  for (Instruction &I : instructions(F)) {
+    if (auto CastOp = dyn_cast<llvm::CastInst>(&I)) {
+      llvm::Type *DstTy = CastOp->getDestTy();
+      auto CastOpcode = CastOp->getOpcode();
+      if ((CastOpcode == llvm::Instruction::FPToUI &&
+           DstTy->getScalarType()->getPrimitiveSizeInBits() <= 32) ||
+          (CastOpcode == llvm::Instruction::FPToSI &&
+           DstTy->getScalarType()->getPrimitiveSizeInBits() < 32)) {
+        IRBuilder<> Builder(&I);
+        llvm::Value *Src = CastOp->getOperand(0);
+        auto TmpTy =
+            llvm::VectorType::get(llvm::Type::getInt32Ty(DstTy->getContext()),
+                                  cast<VectorType>(DstTy)->getNumElements());
+        Src = Builder.CreateFPToSI(Src, TmpTy);
+
+        llvm::Instruction::CastOps TruncOp = llvm::Instruction::Trunc;
+        llvm::Value *NewDst = Builder.CreateCast(TruncOp, Src, DstTy);
+        CastOp->replaceAllUsesWith(NewDst);
+        ESIMDToErases.push_back(CastOp);
+      }
+    }
+
+    auto *CI = dyn_cast<CallInst>(&I);
+    Function *Callee = nullptr;
+    if (!CI || CI->isIndirectCall() || !(Callee = CI->getCalledFunction()))
+      continue;
+    StringRef Name = Callee->getName();
+
+    // See if the Name represents an ESIMD intrinsic and demangle only if it
+    // does.
+    if (!Name.startswith(ESIMD_INTRIN_PREF0))
+      continue;
+    // now skip the digits
+    StringRef Name1 = Name.substr(std::strlen(ESIMD_INTRIN_PREF0));
```

Comment:
I'm also not sure why do you need `Name1` here if you can just re-assign result of `substr` or `drop_while` back to `Name`

Reply:
I'll see if I can simplify